### PR TITLE
misc cleanup

### PR DESCRIPTION
**Pull request by tamird**
Also pass `golint` and `errcheck`.

### CODE REVIEW

**Comment by tamird** [comment]
@soheilhy pimng

File: cmux.go [hunk 1]
```diff
@@ -162,18 +169,18 @@ func (l muxListener) Accept() (c net.Conn, err error) {
 	}
 }
 
-type MuxConn struct {
+type muxConn struct {
```
Comment (soheilhy) [inline cmux.go]: Please revert this part. I need MuxConn to be exported. :-)

Comment (tamird) [inline cmux.go]: Can you elaborate? If I re-export it, it needs a doc comment - I wouldn't know what to write.

File: key.pem [hunk 2]
```diff
@@ -0,0 +1,15 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (soheilhy) [inline key.pem]: I would prefer a go generate here.

File: cmux_test.go [hunk 3]
```diff
@@ -180,8 +220,7 @@ func TestClosed(t *testing.T) {
 	lis := mux.Match(Any()).(muxListener)
 	close(lis.donec)
 	mux.serve(closerConn{})
-	_, err := lis.Accept()
-	if _, ok := err.(errListenerClosed); !ok {
-		t.Errorf("expected errListenerClosed got %v", err)
+	if _, err := lis.Accept(); err != ErrListenerClosed {
```
Comment (soheilhy) [inline cmux_test.go]: Please revert. I want to check the interface, not only one instance.

File: example_recursive_test.go [hunk 4]
```diff
@@ -24,14 +26,18 @@ func recursiveServeHTTP(l net.Listener) {
...
+		}
+	}()
 }
 
 func tlsListener(l net.Listener) net.Listener {
 	// Load certificates.
 	certificate, err := tls.LoadX509KeyPair("cert.pem", "key.pem")
 	if err != nil {
-		log.Fatal(err)
+		log.Panic(err)
```
Comment (soheilhy) [inline example_recursive_test.go]: Why do we need to s/Fatal/Panic/g in the example?

File: matchers.go [hunk 5]
```diff
@@ -148,7 +148,9 @@ func matchHTTP2Field(r io.Reader, name, value string) (matched bool) {
 
 		switch f := f.(type) {
 		case *http2.HeadersFrame:
-			hdec.Write(f.HeaderBlockFragment())
+			if _, err := hdec.Write(f.HeaderBlockFragment()); err != nil {
```
Comment (soheilhy) [inline matchers.go]: Please also send this in a separate commit. :-)

File: example_recursive_test.go (hunk 4 above)
Comment (tamird) [inline example_recursive_test.go]: `log.Fatal` is very opaque; you get a short error message and the program exits. `log.Panic` gives you a stack trace; before this change it was not even possible to determine on which line the examples were failing.

File: cmux_test.go (hunk 3 above)
Comment (tamird) [inline cmux_test.go]: This error is a singleton, and only the instance is exported; this test provides documentation of the behaviour, but the old version does not.

Comment (soheilhy) [inline cmux_test.go]: I know, this is the only instance; but only for now. I have plans to add more of them later on. thanks.

File: cmux.go (hunk 1 above)
Comment (soheilhy) [inline cmux.go]: I'm using it in an personal project which isn't public yet. :-) 

Ah, then please add something simple for now:

// MuxConn wraps a net.Conn and provides transparent sniffing of connection data.

File: example_recursive_test.go (hunk 4 above)
Comment (soheilhy) [inline example_recursive_test.go]: SGTM. Thanks. Please include this either in a separate commit or in your lint commit.

File: key.pem (hunk 2 above)
Comment (soheilhy) [inline key.pem]: Why nondeterministic? Would it matter if you and I get different private keys?

**Comment by soheilhy** [comment]
hehe, some of my comments were in your commit not on the pull request. :))

Thanks for doing this, again!

File: example_recursive_test.go (hunk 4 above)
Comment (shurcooL) [inline example_recursive_test.go]: How does `log.Fatal(err)` compare to/differer from `panic(err)`?

From https://godoc.org/log#Panic:

> Panic is equivalent to Print() followed by a call to panic().

If I'm understanding that right, in this case, they are equivalent, since the only parameter is a single `error` value. Is that right?

Comment (tamird) [inline example_recursive_test.go]: With the standard caveat that https://godoc.org/log#SetOutput can be used to redirect the printed error, yes

Comment (soheilhy) [inline example_recursive_test.go]: Sure, they are different, but since this is just an example, perhaps a simple `panic` will do just fine?

Comment (tamird) [inline example_recursive_test.go]: OK.

Comment (soheilhy) [inline example_recursive_test.go]: Thank you! :-)

File: key.pem (hunk 2 above)
Comment (tamird) [inline key.pem]: `go generate` is not integrated into `go test` et al., so putting this into `go generate` would mean that users cannot run the tests until they create these certificates.

**Comment by tamird** [comment]
ok, PTAL.

**Comment by tamird** [comment]
It's not possible to revert the examples if you still intend to run them. Serving on the socket blocks, so the example would never terminate.

**Comment by soheilhy** [comment]
That's right, and nor does any example in the net/http package. It's fine to have package examples block (I mean that ones that are shown with package "main" by godoc).

And sorry for the late reply, didn't receive an email on this.

**Comment by tamird** [comment]
I'm just going to maintain a fork. Thanks.

**Comment by tamird** [comment]
Feel free to merge all but the last commit if you want.

**Comment by tamird** [comment]
I've removed the last commit.

**Comment by soheilhy** [comment]
Manually edited and merged, except "go tool vet --shadow ." 
thanks for the effort!
